python_ta/checkers/inconsistent_or_missing_returns_checker.py

```diff
@@ -0,0 +1,82 @@
+"""
+Check for inconsistent return statements in functions and missing return statements in non-None functions.
+"""
+
+from typing import Optional
+
+from astroid import nodes
+from pylint.checkers import BaseChecker
+from pylint.lint import PyLinter
+
+from python_ta.cfg import ControlFlowGraph
+
+
+class InconsistentReturnChecker(BaseChecker):
+
+    name = "inconsistent-or-missing-returns"
+    msgs = {
+        # as R1710 is being disabled, we replace it with an identical message
+        "R9710": (
+            """This function has at least one case where the function body will execute without ending in an explicit return statement.
+            You should check your code to make sure every possible execution path through the function body ends in a return statement.
+            Note: one common source of this error is if you're using if statements without an explicit else branch.
+            In this case, you should consider revising your code to either add an else branch,
+            or, if you are confident that the if and elif conditions cover all possible cases,
+            "you can convert the final "elif " into an " else ".",""",
+            "inconsistent-returns",
+            "Used to replace R1710 message inconsistent-return-statements",
+        ),
+        "R9711": (
+            "Missing return statement in function",
+            "missing-return-statements",
+            "Used when a function does not have a return statement and whose return type is not None",
+        ),
+    }
+
+    def __init__(self, linter: Optional[PyLinter] = None) -> None:
+        super().__init__(linter=linter)
+
+    def visit_functiondef(self, node) -> None:
+        """Visit a function definition"""
+        self._check_return_statements(node)
+
+    def _check_return_statements(self, node) -> None:
+        """
+        Construct a CFG from the function. Check for inconsistent returns if there are
+        multiple return statements, and missing return statements if there are none.
+        """
+        if (
+            node.returns is None
```

Comment:
This is a bit subtle. I believe `node.returns is None` is true when there's _no_ type annotation at all? In this case I wonder what pylint's R1710 does.

I suspect that in this case, the check could still execute, but then there's a special case when there's no return statement, or only bare `return`s, in which case no error is reported. That is, in this case the check is conditioned on whether at least one return statement has an expression, in which case all of them should.

Reply:
Yes, `node.returns is None` does mean that there is no return type annotation. Previously I thought if having no annotation is equivalent to having return type as `None`. I guess what you mean here is that when there is no type annotation, we only check for inconsistent returns but not missing returns?

Reply:
@Raine-Yang-UofT right, "no annotation" should be interpreted as "unspecified", not "return type `None`".

Your interpretation is correct: when no type annotation is provided, check for inconsistent returns but not missing returns. However, note that if *all* return statements are just `return`, this should not be considered "inconsistent returns".